agents/dns/dns.py

```diff
@@ -23,9 +27,48 @@ def __init__(
             log_path=log_path,
         )
 
-    def service_healthy(self, timeout=2, payload=None, packet_size=0):
-        # TODO: extends check - send a DNS query and get a valid DNS response.
-        status = super(DNSAgent, self).service_healthy()
-        return status and self.is_port_open(
-            timeout=timeout, payload=payload, packet_size=packet_size
-        )
+    def run_dig(self, query_domain):
+        """
+        Run the dig command to query DNS locally.
+        Returns:
+            tuple: (returncode, stdout, stderr)
+        """
+        process = subprocess.Popen(
+                ['dig', '@localhost', query_domain, '+short'],
+                stdout=subprocess.PIPE,
+                stderr=subprocess.PIPE,
+            )
+        output, error = process.communicate()
+        return process.returncode, output, error
+
+    def is_dns_running(self):
+        """
+        Check if DNS is responding to queries using dig.
+        Returns:
+            bool: True if DNS query succeeds, False otherwise.
+        """
+        try:
+            # Run the dig command
+            returncode, output, error = self.run_dig(query_domain='google.com')
+
+            if returncode != 0:
+                return False
+
+            # If output is not empty and IP valid, DNS is working
+            return bool(output.strip()) and is_valid_ip(output)
+
+        except OSError as e:
+            # Command not found or failed to execute
+            maybe_log_message(
+                'DNS check failed: %s' % e,
+                self.logger,
+                fallback_logger=self.fallback_logger,
+                exc_info=True,
+            )
+            return False
+
+    def service_healthy(self):
+        return (
+            super(DNSAgent, self).service_healthy()
+            and self.is_dns_running()
+            )
```

Comment:
Currently parent's `service_healthy` only runs `_is_process_running`, but not `is_port_open`. Consider calling it explicitly in `DNSAgent` - especially that it now allows sending a dummy package via UDP and checking response.